### PR TITLE
[#653] Refactor Backup flow screens

### PR DESCRIPTION
**Pull request by HonzaR**
This PR brings:
- Using Compose Scaffold component to have consistency across the app and future proof solution
- Moved copy-to-buffer option to Toolbar overflow menu
- Followed design to enable toolbar back navigation button (system back navigation covered too) on several screens 
- Removed Backup ViewModel component and switched screen save/restore to Compose Saver technique 
- Fixed scrolling on several screens (tested on very small devices, like 4 WVGA Nexus S in landscape mode)
- Fixed screenshot tests for the Backup flow screens

Note: This PR is still rather functional than UI styling according to designs (FYI @geffenz). UI styling of the entire onboarding (like paddings, fonts, colors, positions, etc) will follow as part of the epics https://github.com/zcash/secant-android-wallet/issues/43 and https://github.com/zcash/secant-android-wallet/issues/611.

See attached before/after screenshots and video of behavior on very small device. 

Closes https://github.com/zcash/secant-android-wallet/issues/653, https://github.com/zcash/secant-android-wallet/issues/150

screen_bf

[very_small_screen.webm](https://user-images.githubusercontent.com/9017390/202401985-2b579cf3-3f1a-48a6-b288-a8a0c0dd00af.webm)

This code review checklist is intended to serve as a starting point for the author and reviewer, although it may not be appropriate for all types of changes (e.g. fixing a spelling typo in documentation).  For more in-depth discussion of how we think about code review, please see [Code Review Guidelines](../blob/main/docs/CODE_REVIEW_GUIDELINES.md).

# Author

- [x] Self-review: Did you review your own code in GitHub's web interface? _Code often looks different when reviewing the diff in a browser, making it easier to spot potential bugs._
- [x] Automated tests: Did you add appropriate automated tests for any code changes?
- [ ] Manual tests: Did you update the [manual tests](../blob/main/docs/testing/manual_testing) as appropriate? _While we aim for automated testing of the application, some aspects require manual testing. If you had to manually test something during development of this pull request, write those steps down._
- [ ] Code coverage: Did you check the code coverage report for the automated tests?  _While we are not looking for perfect coverage, the tool can point out potential cases that have been missed._ Code coverage can be generated with: `./gradlew check` for Kotlin modules and `./gradlew connectedCheck -PIS_ANDROID_INSTRUMENTATION_TEST_COVERAGE_ENABLED=true` for Android modules.
- [ ] Documentation: Did you update documentation as appropriate? (e.g [README.md](../blob/main/README.md), etc.)
- [x] Run the app: Did you run the app and try the changes?
- [x] Screenshots: Did you provide before and after UI screenshots in the description of this pull request?  _This is only applicable for changes that modify the UI._
- [x] Rebase and squash: Did you pull in the latest changes from the main branch and squash your commits before assigning a reviewer? _Having your code up to date and squashed will make it easier for others to review. Use best judgement when squashing commits, as some changes (such as refactoring) might be easier to review as a separate commit._

# Reviewer

- [ ] Checklist review: Did you go through the code with the [Code Review Guidelines](../blob/main/docs/CODE_REVIEW_GUIDELINES.md) checklist?
- [ ] Ad hoc review: Did you perform an ad hoc review?  _In addition to a first pass using the code review guidelines, do a second pass using your best judgement and experience which may identify additional questions or comments. Research shows that code review is most effective when done in multiple passes, where reviewers look for different things through each pass._
- [ ] Automated tests: Did you review the automated tests?
- [ ] Manual tests: Did you review the manual tests?
- [ ] How is code coverage affected? _We encourage you to compare coverage before and after changes and when possible, leaving it in a better place._
- [ ] Documentation: Did you review Docs, [README.md](../blob/main/README.md), and [Architecture.md](../blob/main/docs/Architecture.md) as appropriate?
- [ ] Run the app: Did you run the app and try the changes? _While the CI server runs the app to look for build failures or crashes, humans running the app are more likely to notice unexpected log messages, UI inconsistencies, or bad output data. Perform this step last, after verifying the code changes are safe to run locally._

### CODE REVIEW

File: ui-lib/src/androidTest/java/co/electriccoin/zcash/ui/screen/backup/integration/BackupActivityTest.kt [hunk 1]
```diff
@@ -34,18 +34,18 @@ class BackupActivityTest : UiTestPrerequisites() {
...
 
-        assertEquals(BackupStage.EducationOverview, testSetup.getStage())
+        assertEquals(BackupStage.EducationOverview.order, testSetup.getStage().order)
 
         composeTestRule.onNodeWithText(getStringResource(R.string.new_wallet_1_button)).also {
             it.performClick()
         }
 
-        assertEquals(BackupStage.EducationRecoveryPhrase, testSetup.getStage())
+        assertEquals(BackupStage.EducationRecoveryPhrase.order, testSetup.getStage().order)
```
Comment (ccjernigan) [inline ui-lib/src/androidTest/java/co/electriccoin/zcash/ui/screen/backup/integration/BackupActivityTest.kt:43]: Does the .equals() comparison not end up working for these?  It seems like that would be more robust and would avoid exposing the `order` value.

**Review by ccjernigan — COMMENTED** [review]
A good set of enhancements to enable future improvements for the onboarding flow!

File: ui-lib/src/main/java/co/electriccoin/zcash/ui/screen/backup/model/BackupStage.kt [hunk 2]
```diff
@@ -3,37 +3,64 @@ package co.electriccoin.zcash.ui.screen.backup.model
...
+        fun values(): Array<BackupStage> = arrayListOf<BackupStage>().apply {
+            EducationOverview.let { add(it.order, it) }
+            EducationRecoveryPhrase.let { add(it.order, it) }
+            Seed.let { add(it.order, it) }
+            Test(Test.TestStage.InProgress).let { add(it.order, it) }
+            Complete.let { add(it.order, it) }
+        }.toTypedArray()
+    }
+
+    object EducationOverview : BackupStage(EDUCATION_OVERVIEW_ORDER)
```
[the remark this answers is not in the document]
Comment (HonzaR) [reply ui-lib/src/main/java/co/electriccoin/zcash/ui/screen/backup/model/BackupStage.kt:27]: Yes, nice improvement. Unfortunately, this feature is available starting from Kotlin `1.8`. See [How to enable data objects](https://kotlinlang.org/docs/whatsnew1720.html#how-to-enable-data-objects). So we can either switch to kotlin version `1.8-Beta` (which brings other requirements to our code), or change kotlin compiler `languageVersion="1.8"` in build conventions. Or we can just wait for it to come to a production version. What do you think?

File: ui-lib/src/main/java/co/electriccoin/zcash/ui/screen/backup/model/BackupStage.kt [hunk 3]
```diff
@@ -3,37 +3,64 @@ package co.electriccoin.zcash.ui.screen.backup.model
...
+        // Note: the order index is used to manage progression through each stage
+        // so be careful if reordering these
+        private const val EDUCATION_OVERVIEW_ORDER = 0
+        private const val EDUCATION_RECOVERY_PHRASE_ORDER = 1
+        private const val SEED_ORDER = 2
+        private const val TEST_ORDER = 3
+        private const val COMPLETE_ORDER = 4
+
+        // This is a rough way how to minimize orders handling
+        fun values(): Array<BackupStage> = arrayListOf<BackupStage>().apply {
```
[the remark this answers is not in the document]
Comment (HonzaR) [reply ui-lib/src/main/java/co/electriccoin/zcash/ui/screen/backup/model/BackupStage.kt:18]: I simplified it a bit with lazy initialization and a less rough way of filling the list.

File: ui-lib/src/main/java/co/electriccoin/zcash/ui/screen/backup/model/BackupStage.kt [hunk 4]
```diff
@@ -3,37 +3,64 @@ package co.electriccoin.zcash.ui.screen.backup.model
...
 
-enum class BackupStage {
-    // Note: the ordinal order is used to manage progression through each stage
-    // so be careful if reordering these
-    EducationOverview,
-    EducationRecoveryPhrase,
-    Seed,
-    Test,
-    Complete;
+sealed class BackupStage(internal val order: Int) {
```
[the remark this answers is not in the document]
Comment (HonzaR) [reply ui-lib/src/main/java/co/electriccoin/zcash/ui/screen/backup/model/BackupStage.kt:6]: Good idea. So as we discussed in our meeting, I flattened the stages structure to have Test and Failure a regular stages. I also introduced the special CheckSeed screen, which can be accessible from Complete stage only. See last commit message for a full changes overview.

File: ui-lib/src/androidTest/java/co/electriccoin/zcash/ui/screen/backup/integration/BackupActivityTest.kt (hunk 1 above)
Comment (HonzaR) [reply ui-lib/src/androidTest/java/co/electriccoin/zcash/ui/screen/backup/integration/BackupActivityTest.kt]: Yes, it works at the end. Thanks.

File: ui-lib/src/main/java/co/electriccoin/zcash/ui/screen/backup/model/BackupStage.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (ccjernigan) [reply ui-lib/src/main/java/co/electriccoin/zcash/ui/screen/backup/model/BackupStage.kt:32]: I don't have a problem using new language features in the Secant app but for this one we can wait until Kotlin 1.8 is released just to keep things simple.

File: ui-lib/src/main/java/co/electriccoin/zcash/ui/screen/backup/view/BackupView.kt [hunk 5]
```diff
@@ -72,32 +89,78 @@ fun BackupWallet(
...
+            is BackupStage.Seed -> SeedPhrase(wallet)
+            is BackupStage.Test -> TestInProgress(
+                selectedTestChoices = choices,
+                onChoicesChanged = onChoicesChanged,
+                splitSeedPhrase = wallet.seedPhrase.split,
+                backupState = backupState
+            )
+            is BackupStage.Failure -> TestFailure()
+            is BackupStage.Complete -> TestComplete()
+            is BackupStage.CheckSeed -> SeedPhrase(wallet)
```
Comment (ccjernigan) [inline ui-lib/src/main/java/co/electriccoin/zcash/ui/screen/backup/view/BackupView.kt:153]: Fine for now, but note that in the future we'll probably copy-paste the SeedPhrase composable to have one complex version showing Google Credential manager choices and a simple version for re-review at the end.

**Review by ccjernigan — APPROVED** [review]
This is a nice set of consistency improvements for our UI implementation, along with some good future-oriented changes to support google credential manager!

**Pull request merged**